samples/net/sockets/big_http_download/src/big_http_download.c

```diff
@@ -35,11 +35,15 @@
 
 /* This URL is parsed in-place, so buffer must be non-const. */
 static char download_url[] =
+#if defined(CONFIG_SAMPLE_BIG_HTTP_DL_URL)
```

Comment:
Won't this condition always be true? The config is added to the sample unconditionally.

Anyway, I think it would be neater to simply add the following line to the `overlay-tls.conf`:
```
CONFIG_SAMPLE_BIG_HTTP_DL_URL="https://www.7-zip.org/a/7z1805.exe"
```
and give up on these nested ifdefs.

Reply:
If we build with Zephyr build system. But that's not the only target and build system supported, this (and other samples I wrote) also work with standard POSIX make.

> Anyway, I think it would be neater to simply add the following line to the overlay-tls.conf:

Yes, the original idea was to make fully configurable via Kconfig. But that also requires configuring hash value, and as https://github.com/zephyrproject-rtos/zephyr/pull/24124#issuecomment-609791549 comment explains, that's blocked on #23917. Hopefully you agree, that moving URLs to Kconfig and leave hashes in .c would only make matters more confusing. (There's also a question of POSIX compat mentioned above, but as TLS support in this sample isn't POSIX-compliant, that can be discounted.)

Reply:
Ok, let me rephrase, perhaps I wasn't clear enough.

My point is, that the current approach breaks the sample with TLS enabled. In Zephyr environment, the string provided by CONFIG_SAMPLE_BIG_HTTP_DL_URL will be used **unconditionally**, so the following will never be evaluated (in Zephyr):
```
#if !defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS)
    "http://archive.ubuntu.com/ubuntu/dists/xenial/main/installer-amd64/current/images/hd-media/vmlinuz";
#else
    "https://www.7-zip.org/a/7z1805.exe";
#endif /* !defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS) */
```

So we need to move the other address to Kconfig anyway (either by overlay or by a proper default in the Kconfig file).

I understand we need `else` for non-Zephyr environments, but it should be the sample default, in this case:
```
"http://archive.ubuntu.com/ubuntu/dists/xenial/main/installer-amd64/current/images/hd-media/vmlinuz"
```

> Hopefully you agree, that moving URLs to Kconfig and leave hashes in .c would only make matters more confusing.

Yes I agree that might be confusing, but isn't this commit exaclty doing that? Anyway, I'd prefer some  (hopefully temporary) confusion over a broken sample.

Reply:
> CONFIG_SAMPLE_BIG_HTTP_DL_URL will be used unconditionally

Got it, thanks. My preference would be to move to the overlay config indeed.

Reply:
The `defined(CONFIG_SAMPLE_BIG_HTTP_DL_URL)` will match even if the value is empty string. So you need to add runtime check to make sure that the string is not empty.

Reply:
> The defined(CONFIG_SAMPLE_BIG_HTTP_DL_URL) will match even if the value is empty string. So you need to add runtime check to make sure that the string is not empty.

There's already such a check. The behavior with empty string is:

~~~
[00:00:02.380,000] <inf> net_config: Initializing network
[00:00:02.380,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:02.380,000] <inf> net_config: Running dhcpv4 client...
Error: Only http: URLs are supported
exit
~~~

Reply:
Ok, great. Also how these #ifdefs are now set means that the settings in #else branch
```
#if !defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS)
    "http://archive.ubuntu.com/ubuntu/dists/xenial/main/installer-amd64/current/images/hd-media/vmlinuz";
#else
    "https://www.7-zip.org/a/7z1805.exe";
#endif /* !defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS) */
```
 are never set, because the defined() only works properly for boolean types. So we could just remove these then.

Reply:
> are never set, because the defined() only works properly for boolean types.

I mentioned that already in a comment to @rlubos before, those settings are used when building without Zephyr Kconfig system, i.e. for native POSIX systems with `make -f Makefile.posix`.

Reply:
Ok, sorry I missed it.